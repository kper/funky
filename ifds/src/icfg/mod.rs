/// the exploded graph representation of IR code
pub mod graph;
/// this module converts the exploded graph into a .tex file for display
pub mod tikz;

/// convert the IR into an ifds instance
pub mod convert;

/// contains the flow functions which be used to build up the graph
pub mod flowfuncs;

/// Keeps node information about the program
pub mod state;

/// The naive implementation for IFDS.
/// This is just a reference implementation for benchmarking.
pub mod naive;

/// Implementation of the original IFDS problem.
pub mod orig;