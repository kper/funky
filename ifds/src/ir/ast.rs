/// IR datastructure for the webassembly module.
#[derive(Debug, Clone)]
pub struct Program {
    pub functions: Vec<Function>
}

/// IR datastructure for the webassembly function.
#[derive(Debug, Clone, Default)]
pub struct Function {
    /// Name of the function
    pub name: String,
    pub(crate) params: Vec<String>,
    pub(crate) definitions: Vec<String>,
    pub(crate) results_len: usize,
    pub(crate) instructions: Vec<Instruction>,
}

type Dest = String;
type Src = String;
type Reg = String;
type Regs = Vec<String>;

/// IR datastructure for abstracted webassembly instructions.
#[derive(Debug, Clone)]
pub enum Instruction {
    Block(String),
    Unop(Dest, Src),
    BinOp(Dest, Src, Src),
    Const(Dest, f64),
    Assign(Dest, Src),
    Jump(String),
    Call(String, Vec<Reg>, Regs),
    Kill(Dest),
    Conditional(Src, Vec<String>),
    Return(Vec<Reg>),
    Table(Vec<String>),
    Phi(Dest, Src, Src),
    CallIndirect(Vec<String>, Vec<Reg>, Vec<Reg>), // names, parameters, dests
    Unknown(Dest), // Value is not known statically
    Store,
}