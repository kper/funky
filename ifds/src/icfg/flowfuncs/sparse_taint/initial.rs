use crate::icfg::tabulation::sparse::defuse::DefUseChain;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintInitialFlowFunction;

impl SparseInitialFlowFunction for SparseTaintInitialFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        init_facts: &Vec<Fact>,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Edge>> {
        debug!("Calling init flow for {} with pc {}", function.name, pc);

        let mut edges = Vec::new();

        // We need the offset, because not every
        // instruction is taintable. For example BLOCK and JUMP
        // have no registers. That's why skip to the next one.
        let offset = 0;
        let instructions = &function.instructions;

        let init_fact = init_facts.get(0).context("Cannot find taut")?.clone();

        debug!("init fact is {:#?}", init_fact);

        loop {
            let pc = pc + offset;
            let instruction = instructions.get(pc).context("Cannot find instr")?;
            debug!("Next instruction is {:?}", instruction);

            match instruction {
                Instruction::Const(dest, _)
                | Instruction::Unknown(dest)
                | Instruction::Assign(dest, _)
                | Instruction::Unop(dest, _)
                | Instruction::BinOp(dest, _, _)
                | Instruction::Phi(dest, _, _)
                | Instruction::Conditional(dest, _)
                | Instruction::Load(dest, _, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Call(_callee, _, dests) => {
                    // When the analysis starts at a Call Instruction,
                    // then the function doesn't get called.
                    // But all destinations get automatically tainted.

                    let before = vec![init_fact.clone()];
                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }
                    }
                    for dest in dests {
                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Store(_src, offset, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let mem = ctx
                            .state
                            .add_memory_var(function.name.clone(), offset.clone() as usize);

                        let after_var = defuse
                            .demand_inclusive(ctx, &function, &mem.name, pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let mut applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Return(_dests) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: var.clone(),
                            });
                        }
                    }
                }
                Instruction::Block(_) | Instruction::Jump(_) => {
                    panic!("Block or Jump as first instruction is not supported");
                }
                _ => {}
            }

            break;
        }

        Ok(edges)
    }
}
