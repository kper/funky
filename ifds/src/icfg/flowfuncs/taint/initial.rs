use crate::icfg::flowfuncs::*;

pub struct TaintInitialFlowFunction;

impl InitialFlowFunction for TaintInitialFlowFunction {
    fn flow(
        &self,
        function: &AstFunction,
        graph: &mut Graph,
        pc: usize,
        init_facts: &Vec<Fact>,
        normal_flows_debug: &mut Vec<Edge>,
        state: &mut State,
    ) -> Result<Vec<Edge>> {
        debug!("Calling init flow for {} with pc {}", function.name, pc);

        let mut edges = Vec::new();

        // We need the offset, because not every
        // instruction is taintable. For example BLOCK and JUMP
        // have no registers. That's why skip to the next one.
        let mut offset = 0;
        let instructions = &function.instructions;

        let mut init_fact = init_facts.get(0).context("Cannot find taut")?.clone();

        loop {
            let pc = pc + offset;
            let instruction = instructions.get(pc).context("Cannot find instr")?;
            debug!("Next instruction is {:?}", instruction);

            state.add_statement(
                function,
                format!("{:?}", instruction),
                pc,
                &"taut".to_string(),
            )?;

            let _after2 = state
                .get_facts_at(&function.name, pc)?
                .filter(|x| x.belongs_to_var == "taut".to_string())
                .collect::<Vec<_>>();

            let after2 = _after2.get(0).unwrap();

            edges.push(Edge::Path {
                from: init_fact.clone().clone(),
                to: after2.clone().clone(),
            });

            match instruction {
                Instruction::Const(reg, _) => {
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, reg)?;

                    let before2 = vec![init_fact.clone()];

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == reg)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a.clone(),
                        });
                    }
                }
                Instruction::Assign(dest, _src) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == dest)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                Instruction::Unop(dest, _src) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == dest)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                Instruction::BinOp(dest, _src, _src2) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == dest)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                Instruction::Conditional(reg, _) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, reg)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| x.belongs_to_var == "taut".to_string())
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                Instruction::Block(_) | Instruction::Jump(_) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(
                        function,
                        format!("{:?}", instruction),
                        pc + 1,
                        &"taut".to_string(),
                    )?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| x.belongs_to_var == "taut".to_string())
                        .cloned();

                    // Replace the init_fact for the next iteration.
                    // Because, we would skip one row if not.
                    let after3 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| x.belongs_to_var == "taut".to_string())
                        .cloned();

                    init_fact = after3.collect::<Vec<_>>().get(0).unwrap().clone();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }

                    offset += 1;

                    continue;
                }
                Instruction::Phi(dest, _src1, _src2) => {
                    let before2 = vec![init_fact.clone()];

                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                    let after2 = state
                        .get_facts_at(&function.name, pc + 1)?
                        .filter(|x| &x.belongs_to_var == dest)
                        .cloned();

                    for (b, a) in before2.into_iter().zip(after2) {
                        normal_flows_debug.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                        edges.push(Edge::Path {
                            from: init_fact.clone().clone(),
                            to: a,
                        });
                    }
                }
                _ => {}
            }

            break;
        }

        Ok(edges)
    }
}
