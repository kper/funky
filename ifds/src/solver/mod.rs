//! The implementation for extracting the taints from the graph.

use crate::icfg::graph::{Edge, Graph};
use anyhow::{Context, Result};
use std::collections::HashSet;

type PC = usize;

pub struct IfdsSolver;

/// Represents the taints from the execution of the solver.
#[derive(Debug)]
pub struct Taint {
    pub variable: String,
    pub function: String,
    pub pc: PC,
}

/// Request is the datastructure which defines at what instruction the [`Graph`] should be build from.
#[derive(Debug, Clone)]
pub struct Request {
    pub variable: Option<String>,
    pub function: String,
    pub pc: PC,
}

pub trait Solver {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>>;

    /// Check if the statement at `req` and the statement `resp` have a taint.
    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool>;

    /// Return all tainted variables names at any statement for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>>;
}

impl Solver for IfdsSolver {
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints: Vec<_> = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    //&& x.get_from().var_is_taut
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| Taint {
                function: x.function.clone(),
                pc: x.next_pc.checked_sub(1).unwrap_or(0),
                variable: x.belongs_to_var.clone(),
            })
            .collect();

        Ok(taints)
    }

    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| x.belongs_to_var.clone())
            .collect::<HashSet<_>>();

        Ok(taints)
    }

    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool> {
        let all_sinks = self
            .all_sinks(graph, req)
            .context("Cannot compute the sinks")?;

        let var = resp
            .variable
            .as_ref()
            .context("Please specify a variable for the response")?;

        Ok(all_sinks
            .into_iter()
            .find(|x| x.function == resp.function && x.pc == resp.pc && &x.variable == var)
            .is_some())
    }
}

pub trait GraphReachability {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Vec<Taint>;
}
