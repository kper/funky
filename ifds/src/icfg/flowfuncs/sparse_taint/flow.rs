use crate::icfg::state::State;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintNormalFlowFunction;

impl SparseTaintNormalFlowFunction {
    pub fn identity<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        fact: &Fact,
        pc: usize,
        defuse: &mut DefUseChain,
        edges: &mut Vec<Edge>,
    ) -> Result<()> {
        let after = defuse.get_next(ctx, function, &fact.belongs_to_var, pc)?;

        for to in after {
            edges.push(Edge::Normal {
                from: fact.apply(),
                to: to,
                curved: false,
            });
        }

        Ok(())
    }
}

impl SparseNormalFlowFunction for SparseTaintNormalFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        variable: &String,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Fact>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut facts = Vec::new();

        let instructions = &function.instructions;
        let instruction = instructions.get(pc);

        if instruction.is_none() {
            debug!("Instruction is none");
            return Ok(Vec::new());
        }

        let instruction = instruction.unwrap();

        debug!("Next instruction is {:?} for {}", instruction, variable);

        let mut nodes = defuse.demand(ctx, function, variable, pc)?;

        match instruction {
            Instruction::Unop(dest, ..) |
            Instruction::Phi(dest, ..) |
            Instruction::BinOp(dest, ..) | 
            Instruction::Assign(dest, ..) => {
                let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                nodes.extend(x.into_iter().map(|x| x.clone()));
            }
            _ => {}
        }
        let nodes = nodes
            .into_iter()
            .filter_map(|x| {
                if x.pc != x.next_pc {
                    Some(x.apply())
                } else {
                    Some(x)
                }
            })
            .collect::<Vec<_>>();

        facts.extend(nodes.into_iter());

        debug!("Out {:#?}", facts);

        Ok(facts)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::icfg::state::State;
    use crate::ir::ast::Program;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_binop() {
        /*
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0
        */

        let func_name = "main".to_string();
        let function = AstFunction {
            name: func_name.clone(),
            definitions: vec![
                "%0".to_string(),
                "%1".to_string(),
                "%2".to_string(),
                "%3".to_string(),
            ],
            instructions: vec![
                Instruction::Const("%0".to_string(), 1.0),
                Instruction::Const("%1".to_string(), 1.0),
                Instruction::BinOp("%2".to_string(), "%0".to_string(), "%1".to_string()),
                Instruction::BinOp("%3".to_string(), "%0".to_string(), "%1".to_string()),
            ],
            ..Default::default()
        };

        let mut graph = Graph::default();
        let mut state = State::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            prog: &Program {
                functions: vec![function.clone()],
            },
            block_resolver: HashMap::default(),
        };

        let pc = 0;

        // fullfilling precondition of `chain.cache()`
        ctx.state.init_function(&function, pc).unwrap();

        let mut defuse = DefUseChain::default();

        let sparse = SparseTaintNormalFlowFunction;

        let facts = sparse
            .flow(&mut ctx, &function, 2, &"%0".to_string(), &mut defuse)
            .unwrap();

        debug!("facts {:#?}", facts);
        assert_eq!(4, facts.len());

        let cmp_facts: Vec<Fact> = vec![
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 3,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
        ];
        assert_eq!(cmp_facts, facts);
    }
}
