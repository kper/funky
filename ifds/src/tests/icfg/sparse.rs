use crate::icfg::tikz2::render_to;
use crate::icfg::{state::State, tabulation::sparse::TabulationSparse};
use crate::solver::Request;
use insta::assert_snapshot;
use log::error;

use crate::grammar::*;

use crate::icfg::flowfuncs::sparse_taint::flow::SparseTaintNormalFlowFunction;
use crate::icfg::flowfuncs::sparse_taint::initial::SparseTaintInitialFlowFunction;
use crate::ir::ast::Function as AstFunction;

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {{
        let mut convert = TabulationSparse::new(
            SparseTaintInitialFlowFunction,
            SparseTaintNormalFlowFunction,
        );

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let res = convert.visit(&prog, &$req);

        if let Err(err) = res {
            error!("ERROR: {}", err);
            err.chain()
                .skip(1)
                .for_each(|cause| error!("because: {}", cause));
            panic!("")
        }

        let (graph, state) = res.unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("{}_dot", $name), output);

        convert
    }};
}

#[test]
fn test_ir_const() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_const",
        req,
        "
         define test (result 0) (define %0) {
            %0 = 1
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_double_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_double_const",
        req,
        "
         define test (result 0) (define %0 %1 %2){
            %0 = 1
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_simple_store() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_store",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            STORE FROM %0 OFFSET 0 + %0 ALIGN 2 32
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_simple_load() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_load",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            %1 = LOAD OFFSET 0 + %0 ALIGN 0
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_chain_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_chain_assign",
        req,
        "
         define test (result 0) (define %0 %1 %2 %3){
            %0 = 1
            %1 = 1
            %2 = %0
            %3 = %2
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_unop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_unop",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = op %0
            %1 = op %0   
        };"
    );
}

#[test]
fn test_ir_binop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_binop",
        req,
        "define test (result 0) (define %0 %1 %2 %3) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_binop_offset() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_binop_offset",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[ignore]
#[test]
fn test_ir_phi() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_phi",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = phi %0 %1
        };"
    );
}

#[ignore]
#[test]
fn test_ir_killing_op() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_killing_op",
        req,
        "define test (result 0) (define %0 %1 %2)  {
            %0 = 1
            %1 = 1
            KILL %0
            KILL %1
            %2 = 1
        };"
    );
}

#[test]
fn test_ir_if_else_binop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    let tabulation = ir!(
        "test_ir_if_else_binop",
        req,
        "define test (result 0) (define %0 %1 %2 %3) {
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = %0
            %2 = 3
            GOTO 3
            BLOCK 2
            %1 = 1
            GOTO 3
            BLOCK 3
            %3 = %1 op %0   
        };"
    );

    let scfg = tabulation
        .get_scfg_graph(&"test".to_string(), &"%1".to_string())
        .unwrap();

    let output = scfg
        .edges
        .iter()
        .map(|x| format!("({}) -> ({})", x.get_from().pc, x.to().pc))
        .collect::<Vec<_>>().join("\n");

    assert_snapshot!(format!("{}_scfg_dot", "test_ir_if_else_binop"), output);
}

#[test]
fn test_ir_if_else() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_if_else",
        req,
        "define main (result 0) (define %0 %1 %2 %3) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = 2
            %2 = 3
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            %3 = %2
        };
        "
    );
}

#[test]
fn test_ir_if() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_if",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 0
            %1 = 2
            %2 = 3
        };
        "
    );
}

#[test]
fn test_ir_loop() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_loop",
        req,
        "define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            %1 = 2
            GOTO 0 
        };
        "
    );
}

#[test]
fn test_ir_table() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_table",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            %1 = 2
            %2 = 3
            BLOCK 1
            %1 = 2
            %2 = 3
            BLOCK 2
            %2 = 4
            TABLE GOTO 0 1 2 ELSE GOTO 2
        };
        "
    );
}
