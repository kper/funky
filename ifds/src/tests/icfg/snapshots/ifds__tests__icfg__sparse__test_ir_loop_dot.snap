---
source: src/tests/icfg/sparse.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (0) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (1) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (2) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (3) at (1, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (4) at (1, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (5) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (6) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (7) at (1, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (8) at (1, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (9) at (1, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (10) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (11) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (12) at (1, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (13) at (1, 2) { };
\node[font=\tiny] (note_0) at (-1.5, 1.5) {Const(0, 1.0)};
\node[font=\tiny] (note_1) at (-1.5, 2.5) {Const(1, 2.0)};
\node[font=\tiny] (note_2) at (-1.5, 3.5) {Jump(0)};
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->, blue] (0) [bend right] edge  node { } (1);
		\path[->, blue] (0) [bend right] edge  node { } (2);
		\path[->, blue] (0) [bend right] edge  node { } (3);
		\path[->, blue] (0) [bend right] edge  node { } (4);
		\path[->, blue] (0) [bend right] edge  node { } (5);
		\path[->, blue] (0) [bend right] edge  node { } (6);
		\path[->, blue] (0) [bend right] edge  node { } (7);
		\path[->, blue] (0) [bend right] edge  node { } (8);
		\path[->, blue] (0) [bend right] edge  node { } (9);
		\path[->] (0) edge (1);
		\path[->] (0) edge (2);
		\path[->] (0) edge (3);
		\path[->] (0) edge (4);


                \end{tikzpicture}
            \end{document}
